src/basic/alloc-util.h

```diff
@@ -184,17 +185,22 @@ void* greedy_realloc0(void **p, size_t need, size_t size);
 #  define msan_unpoison(r, s)
 #endif
 
-/* This returns the number of usable bytes in a malloc()ed region as per malloc_usable_size(), in a way that
- * is compatible with _FORTIFY_SOURCES. If _FORTIFY_SOURCES is used many memory operations will take the
- * object size as returned by __builtin_object_size() into account. Hence, let's return the smaller size of
- * malloc_usable_size() and __builtin_object_size() here, so that we definitely operate in safe territory by
- * both the compiler's and libc's standards. Note that __builtin_object_size() evaluates to SIZE_MAX if the
- * size cannot be determined, hence the MIN() expression should be safe with dynamically sized memory,
- * too. Moreover, when NULL is passed malloc_usable_size() is documented to return zero, and
- * __builtin_object_size() returns SIZE_MAX too, hence we also return a sensible value of 0 in this corner
- * case. */
-#define MALLOC_SIZEOF_SAFE(x) \
-        MIN(malloc_usable_size(x), __builtin_object_size(x, 0))
+void *expand_to_usable(void *p, size_t newsize) _alloc_ (2);
+
+/* This returns the number of usable bytes in a malloc()ed region as per malloc_usable_size(), which may
+ * return a value larger than the size that was actually allocated. Access to that additional memory is
+ * safe but discouraged because it violates the C standard; a compiler cannot see that this as valid. To help
+ * the compiler out, the MALLOC_SIZEOF_SAFE macro 'allocates' the usable size using a dummy allocator
+ * function expand_to_usable. */
+#define MALLOC_SIZEOF_SAFE(x)                                           \
+        ({                                                              \
+                size_t full_size = malloc_usable_size((x));             \
+                void *newx _unused_;                                    \
+                __builtin_choose_expr(__builtin_constant_p((x)),        \
+                                      newx, (x))                        \
+                        = expand_to_usable((x), full_size);             \
```

Comment:
why do we need the assignment? and why do we need the __builtin_choose_expr() stuff? i mean, we could just do the `newx` branch in all cases at the very least?

i mean, under the assumption is is to trick out the compiler to not generate warnings that the allocation result is unused, then assiging it always to newx should be safe, no?

Reply:
I had to do this to trick `test-alloc-util.c` which passes the constant `(float) NULL`, which cannot be an lvalue. I could fix up the test instead to put `NULL` into a `float *` so that it works and we then get rid of the `__builtin_choose_expr`.  Would that be OK?

Reply:
Specifically, these tests:

```
        assert_se(MALLOC_ELEMENTSOF((float*) NULL) == 0);
        assert_se(MALLOC_SIZEOF_SAFE((float*) NULL) == 0);
```

Reply:
OK I think I understood your suggestion now. Assigning to newx is pointless because the compiler has to see the input pointer being updated. So `x` needs to eventually be updated to point to the 'reallocated' memory.

Reply:
for compat with that test, maybe just special case NULL. The whole excercise is pointless for NULL, so you might as well shortcut it. And it's the only case where passing a constant expression would ever be OK and safe.

Reply:
We'd still need the __builtin_choose_expr hack with the additional NULL check because that's the only way to have the NULL not evaluated as an lvalue.

That said, there's really no reason to call malloc_usable_size with anything that is not a pointer returned by malloc, can't we just adjust the test like so:

```
        float *nullptr = NULL;
        assert_se(MALLOC_ELEMENTSOF(nullptr) == 0);
        assert_se(MALLOC_SIZEOF_SAFE(nullptr) == 0);
```

That would also make `x` safe to expand multiple times, since we need to do that to update it.

Reply:
Actually scratch the safety bit; a `*bufp++` for example could be a valid lvalue won't be safe to evaluate twice.